### PR TITLE
212 delete post

### LINKED ISSUE #212: [212] Admin Delete Post
**Current state**
There is no admin role or any way to remove posts

**Expected state**
Create admin role in database and make admins able to soft-delete a post (deleted_by which stores admin id)

**TO DO**
- [ ] Task 1 ?
- [ ] Task 2 ?
- [ ] Task 3 ?

**Necessities**
#187 

**Branch**
212-admin-delete-post

### PR DESCRIPTION
**Pull request by ryankroon00**
closes #212 

<br />

 **What is it supposed to do**  
This pull request makes sure a mod, admin and head admin are able to archive posts and if a post is archived it isn't displayed.

<br />

 **How can we test your branch**  
- go to the backend folder
- run "npm run reset"
- login as admin 
- go to a post
- click on the postbanner
- click confirm
- the post shouldn't be displayed on the homepage, profile page

<br />

**1. General checklist**
- [x] a. Tested happy flow.
- [x] b. Tested unhappy flow.
- [x] c. No unexpected exceptions.
- [x] d. No code with high complexity. (Big O)
- [x] e. EOF newline.
- [x] f. No floating comma's and spaces.
- [x] g. No unexpected/unhandled errors.
- [ ] h. Lines no longer than 'guideline' in IDE.
- [x] i. Correct indentation.
- [x] j. No usage of legacy functionalities.
- [x] k. (JS) Curly braces on the sameline as reserved keywords *(`if () {` instead of `{` on line below)*
- [x] l. (JS) No double quotes.
- [x] m. (JS) Single newline above unique code structures and reserved keywords.
- [x] n. (JS) Variables bundled where possible.
- [x] o. (JS) No semi-colons except where they have a functional necessity, such as for-loops.
- [x] p. (JS) No code that is meant exclusively for development, such as console.logs for data preview.

<br />

**2. If relevant, front-end checklist**
- [x] a. Responsive. (Desktop all the way to mobile 320px)
- [x] b. Using SCSS variables where relevant.
- [x] c. Relevant class names
- [x] d. No modified global SCSS properties, except if necessary and fully tested the impact.
- [ ] e. Checked [caniuse](https://caniuse.com) for more modern CSS properties.

<br />

**3. If relevant, back-end checklist**
- [x] a. No unhandled promise rejections.
- [x] b. Seeders still functional.
- [x] c. No primary-key violations.
- [x] d. Normalization, atleast [3NF](https://en.wikipedia.org/wiki/Database_normalization).
- [x] e. Used guards (middleware) where necessary and tested if route is guarded.

<br />

**4. If relevant, test these browsers**
- [x] a. Chrome
- [x] b. Firefox
- [ ] c. Opera
- [ ] d. Edge
- [ ] e. Safari

---

**PR Rules**  
*1. Add relevant labels.*  
*2. Select atleast two reviewers.*  
*3. Assign all reviewers allowed to merge the branch, including yourself.*  
*4. If relevant to the sprint, put related issue in 'Awaiting Approval' column on project board.*  
*5. Add to the corresponding [milestone](https://github.com/AjobK/Seaqull/milestones).*

<br />

**Reviewer info**  
*1. You can use aliases to refer to specific PR checklist items. For example 'Hey @JohnDoe, for this line please check 3.a and 1.d.'*  
*2. Where possible, please use the suggestions feature in github so the author has a better idea of the solution you had in mind.*   

<br />

> This form should be filled in by the author of this PR.

### CODE REVIEW

**Comment by MrGooss** [comment]
Fix conflicts please bro, it's been a week....

File: backend/app/controllers/profileController.ts [hunk 1]
```diff
@@ -321,10 +321,31 @@ class ProfileController {
     }
 
     private checkPasswordStrength(password: string): string {
-        if (!isStrongPassword(password, { minSymbols: 0 })) {
-            return 'Password is too weak.\nUse lowercase letter(s), uppercase letter(s) and number(s).\nShould be atleast 8 characters long.'
+        let error = ''
+
+        if (!isLength(password, { min: 8, max: 20 })) {
+            error = error + 'atleast 8 characters'
```
Comment (MrGooss) [inline backend/app/controllers/profileController.ts]: space between "atleast" it's "at least"

File: backend/app/controllers/profileController.ts [hunk 2]
```diff
@@ -321,10 +321,31 @@ class ProfileController {
...
+        if (password.search(/[A-Z]/) < 1 && password.search(/[a-z]/) < 1) {
+            const warningCasing = 'lowercase and uppercase letters'
+
+            if (error != '') error = error + ', '
+
+            error = error + warningCasing
+        }
+
+        if (password.search(/\d/) < 1) {
+            const warningNumber = 'use a number'
```
Comment (MrGooss) [inline backend/app/controllers/profileController.ts]: "use at least one number"

File: backend/app/controllers/profileController.ts [hunk 3]
```diff
@@ -321,10 +321,31 @@ class ProfileController {
     }
 
     private checkPasswordStrength(password: string): string {
```
Comment (MrGooss) [inline backend/app/controllers/profileController.ts]: Validator is not showing on user registration, this should be fixed in another branch

File: frontend/src/components/postBanner/index.js [hunk 4]
```diff
@@ -23,16 +53,24 @@ class PostBanner extends Component {
...
+        <div className={styles.innerWrapper}>
+          <div className={styles.info}>
+            <Link to={`/profile/${author.name}`} className={styles.profileLink}>
+              <div className={styles.infoInner}>
+                <div className={styles.picture} style={{ backgroundImage: `url(${author.avatarURL || ''})` }} />
+                <div className={styles.user_info}>
+                  <h2 className={[styles.name].join(' ')}>{ author.name || ''}</h2>
+                  <div className={styles.achieved}>
+                    <span className={styles.level}>{ Number.isSafeInteger(author.level) ? author.level : '-1'}</span>
+                    <h3 className={styles.role}>{ author.title || ''}</h3>
```
Comment (MrGooss) [inline frontend/src/components/postBanner/index.js]: Spaces between curly braces on classnames

File: frontend/src/pages/post/index.js [hunk 5]
```diff
@@ -164,7 +179,7 @@ class Post extends App {
 
         return (
             <Standard className={[styles.stdBgWhite]}>
-                <PostBanner author={isOwner ? ownerAuthor : author} isOwner={isOwner} />
+                <PostBanner archivePost={this.archivePost.bind(this)} author={author} isOwner={isOwner} />
```
Comment (MrGooss) [inline frontend/src/pages/post/index.js]: Spaces between curly braces

**Review by MrGooss — CHANGES_REQUESTED** [review]
Found some issues regarding spacing, also some issues belonging to another branch

File: backend/app/controllers/postController.ts [hunk 6]
```diff
@@ -224,6 +224,20 @@ class PostController {
...
+    public archivePost = async (req: Request, res: Response): Promise<any> => {
+        const { path } = req.body
+        const post = await this.dao.getPostByPath(path)
+
+        if (!post) res.status(404).json({ error: 'Not found' })
+
+        post.archived_at = Date.now()
+
+        const newPost = await this.dao.updatePost(post)
+        if (!newPost) return res.status(500).json({ error: 'Could not archive post' })
```
Comment (AjobK) [inline backend/app/controllers/postController.ts:287]:
```suggestion
        
        if (!newPost) return res.status(500).json({ error: 'Could not archive post' })
```

Newline

**Comment by AjobK** [comment]
Please capitalize button words. ('Confirm' instead of 'confirm')

image

**Comment by AjobK** [comment]
Able to click on delete post when logged out? When logged in on an account that shouldn't be able to delete posts I can't see it though which is good.

image

**Comment by AjobK** [comment]
Post removal should be stored in a post remove database table. Store staff_id, post_id and archived_at

**Review by AjobK — CHANGES_REQUESTED** [review]
Check previous comments and this minor thing

**Review by AjobK — APPROVED** [review]
Looks good. We will fix the 'Not loaded' to '404' in a different branch.

**Pull request merged**